### ISSUE TITLE
Add Active Stack as part of configuration YAML

### ISSUE THREAD

**Issue by strickvl** [zenml-io/zenml#2261]
## Open Source Contributors Welcomed!
Please comment below if you would like to work on this issue!

### Contact Details [Optional]

support@zenml.io

## What happened?
After a discussion about the Run Config YAML and potential improvements, it became apparent that including the active stack name or ID in the configuration YAML would significantly enhance usability. This change would allow users to specify which stack the pipeline should run on directly within the config YAML.

## Task Description
Update the ZenML configuration YAML structure to include an option for setting the active stack. Users should be able to specify the stack name or ID under a dedicated section in the YAML, which ZenML will then use to determine the active stack for pipeline execution.

## Expected Outcome
- The configuration YAML should have a clear and user-friendly section for specifying the active stack.
- ZenML should correctly interpret and use the specified stack for running the pipeline.
- This update will provide users with greater control and flexibility in their pipeline execution environment.

## Steps to Implement
- Design and implement a new section in the Run Config YAML format for specifying the active stack (name or ID).
- Update the ZenML backend to read and apply this configuration when initializing pipeline runs.
- Ensure backward compatibility and graceful handling when the stack section is not specified in existing YAMLs.
- Test the implementation thoroughly to ensure the correct stack is used based on the YAML configuration. (Include tests in PR).
- Update documentation to guide users on how to use this new feature in their YAML configurations.

## Additional Context
This enhancement is part of ongoing efforts to make ZenML more flexible and user-centric, enabling users to tailor their MLOps workflows more precisely to their needs.

cc: @schustmi @znegrin

## Code of Conduct
- [ ] I agree to follow this project's Code of Conduct

**Comment by prasadshreyas** [comment]
Hi @strickvl, I'm interested in working on this

**Comment by kabinja** [comment]
Hello, if that task is still available, I can start working on it :)

**Comment by strickvl** [comment]
@kabinja yes for sure! let us know if you have any questions, and please do [follow the CONTRIBUTING document](https://github.com/zenml-io/zenml/blob/main/CONTRIBUTING.md), esp w/r/t to branching and making your PR off the `develop` branch!

**Comment by schustmi** [comment]
@kabinja One little detail for you that will be quite helpful: [This](https://github.com/zenml-io/zenml/blob/d7e58bcd65bfac25a9047e959831a5f0d7c3db9d/src/zenml/config/pipeline_run_configuration.py#L27) is the pydantic class that defines the structure of the yaml configuration file.

**Comment by rahuja23** [comment]
Is this issue already resolved?

**Comment by htahir1** [comment]
@rahuja23 no we're still waiting for some activity here.. I think @kabinja seems to have moved on from the task. If not, happy to pick it up from here / make another PR for it

**Comment by htahir1** [comment]
There is a PR here that is stuck for the last 3 weeks https://github.com/zenml-io/zenml/pull/2370#issuecomment-2217234060

**Comment by rahuja23** [comment]
So I can contribute on this issue then?

**Comment by htahir1** [comment]
@rahuja23 yes i would say so, there is a PR already but its stale. You can start from there or from scratch

**Comment by rahuja23** [comment]
@htahir1  Perfect thanks!

**Comment by kabinja** [comment]
Hello, sorry for not getting back earlier. I won't have much time this summer so please go ahead. If you have some questions about what I tried to do, please do not hesitate to reach out to me.

**Comment by preyasshah9** [comment]
Hi @htahir1,

I'm an experienced software developer but new to open source contribution. Are these the active things still required to be done to resolve this issue? 
https://github.com/zenml-io/zenml/pull/2370#issuecomment-2217234060

I would like to take a shot at handling these comments, if that sounds good to you!

**Comment by schustmi** [comment]
@preyasshah9 Yep all those points are still relevant. I'll add some more explanation here:
- `zenml.cli.utils.temporary_active_stack`: We already have a context manager that temporarily sets a default stack. This should probably be moved out of the CLI utilities into something more generic (`zenml.utils.___`) if we want to reuse outside of the CLI context.
- This class `zenml.config.pipeline_run_configuration.PipelineRunConfiguration` defines the options that can be configured using the yaml file.
- You'll need to adjust `zenml.pipelines.run_utils.validate_run_config_is_runnable_from_server(...)` and throw an error if the stack option is set. This is necessary as switching stacks when running a template from the server is currently not supported.

**Comment by preyasshah9** [comment]
@schustmi I have created the Draft PR
https://github.com/zenml-io/zenml/pull/3372

Let me know if this summarizes the requests you have laid out!

**Comment by bcdurak** [comment]
Hey @preyasshah9, I have also left a comment on the PR. I'm closing this due to staleness. Feel free to reopen it if you would like to work on it again.
